### ISSUE TITLE
Very high cpu on mongod process when mongo-connector is running

### ISSUE THREAD

**Issue by ryan-lang** [mongodb-labs/mongo-connector#329]
This could very well be an issue that doesn't involve mongo-connector, but I'm not sure where else to start...

I have a 3-member replica set, with each member running on a separate machine.  I have mongo-connector running on a 4th machine.

While mongo-connector is running, the CPU on my rs primary spikes immediately, and averages 50-75%, compared to the 5-10% when mongo-connector is not running.

Nothing seems out of the ordinary in the connector logs and the instance that mongod is running on should be plenty big to handle normal loads.

Any help debugging this would be much appreciated.  Ideally I'd like to run another connector in parallel, but that's out of the question until I can figure this out...

**Comment by ryan-lang** [comment]
In case it's helpful, a sample of my connector log:
```
2015-10-01 17:49:47,793 [DEBUG] mongo_connector.oplog_manager:664 - OplogThread: reading last checkpoint as Timestamp(1443702541, 8) 
2015-10-01 17:49:47,793 [DEBUG] mongo_connector.oplog_manager:648 - OplogThread: oplog checkpoint updated to Timestamp(1443702541, 8)
2015-10-01 17:49:48,928 [DEBUG] mongo_connector.oplog_manager:167 - OplogThread: Last entry is the one we already processed.  Up to date.  Sleeping.
2015-10-01 17:49:49,929 [DEBUG] mongo_connector.oplog_manager:155 - OplogThread: Getting cursor
2015-10-01 17:49:49,930 [DEBUG] mongo_connector.oplog_manager:664 - OplogThread: reading last checkpoint as Timestamp(1443702541, 8) 
2015-10-01 17:49:49,930 [DEBUG] mongo_connector.oplog_manager:648 - OplogThread: oplog checkpoint updated to Timestamp(1443702541, 8)
2015-10-01 17:49:51,072 [DEBUG] mongo_connector.oplog_manager:167 - OplogThread: Last entry is the one we already processed.  Up to date.  Sleeping.
2015-10-01 17:49:52,073 [DEBUG] mongo_connector.oplog_manager:155 - OplogThread: Getting cursor
2015-10-01 17:49:52,073 [DEBUG] mongo_connector.oplog_manager:664 - OplogThread: reading last checkpoint as Timestamp(1443702541, 8) 
2015-10-01 17:49:52,073 [DEBUG] mongo_connector.oplog_manager:648 - OplogThread: oplog checkpoint updated to Timestamp(1443702541, 8)
2015-10-01 17:49:53,223 [DEBUG] mongo_connector.oplog_manager:167 - OplogThread: Last entry is the one we already processed.  Up to date.  Sleeping.
2015-10-01 17:49:54,224 [DEBUG] mongo_connector.oplog_manager:155 - OplogThread: Getting cursor
2015-10-01 17:49:54,224 [DEBUG] mongo_connector.oplog_manager:664 - OplogThread: reading last checkpoint as Timestamp(1443702541, 8) 
2015-10-01 17:49:54,224 [DEBUG] mongo_connector.oplog_manager:648 - OplogThread: oplog checkpoint updated to Timestamp(1443702541, 8)
2015-10-01 17:49:55,373 [DEBUG] mongo_connector.oplog_manager:167 - OplogThread: Last entry is the one we already processed.  Up to date.  Sleeping.
2015-10-01 17:49:56,374 [DEBUG] mongo_connector.oplog_manager:155 - OplogThread: Getting cursor
2015-10-01 17:49:56,374 [DEBUG] mongo_connector.oplog_manager:664 - OplogThread: reading last checkpoint as Timestamp(1443702541, 8) 
2015-10-01 17:49:56,374 [DEBUG] mongo_connector.oplog_manager:648 - OplogThread: oplog checkpoint updated to Timestamp(1443702541, 8)
2015-10-01 17:49:57,513 [DEBUG] mongo_connector.oplog_manager:167 - OplogThread: Last entry is the one we already processed.  Up to date.  Sleeping.
2015-10-01 17:49:58,514 [DEBUG] mongo_connector.oplog_manager:155 - OplogThread: Getting cursor
2015-10-01 17:49:58,514 [DEBUG] mongo_connector.oplog_manager:664 - OplogThread: reading last checkpoint as Timestamp(1443702541, 8) 
2015-10-01 17:49:58,515 [DEBUG] mongo_connector.oplog_manager:648 - OplogThread: oplog checkpoint updated to Timestamp(1443702541, 8)
2015-10-01 17:49:59,648 [DEBUG] mongo_connector.oplog_manager:167 - OplogThread: Last entry is the one we already processed.  Up to date.  Sleeping.
2015-10-01 17:50:00,649 [DEBUG] mongo_connector.oplog_manager:155 - OplogThread: Getting cursor
```

**Comment by behackett** [comment]
If the source mongo instance has a low volume of changes, you could be hitting https://jira.mongodb.org/browse/SERVER-10226

**Comment by ryan-lang** [comment]
Interesting - I could certainly see that being the case. I will try to confirm before I close this issue. Thanks!

**Comment by behackett** [comment]
This problem should be resolved once and for all with MongoDB 3.2, but there will also have to be an update to mongo-connector to use PyMongo >= 3.2 when it comes out.

**Comment by phber** [comment]
I am getting identical issues on a 1-member replica set running Mongo 3.0.6. I did not fully understand the reason for the problem, but will upgrading to 3.2 solve my issues or is there any other fix?

**Comment by behackett** [comment]
The server side problem is solved in MongoDB 3.2, but to take advantage of that fix mongo-connector has to use PyMongo 3.0+. That work is in code review now, and will be available in the next release.

**Comment by phber** [comment]
So simply upgrading to MongoDb 3.2 is not enough to fix it? Is there a temporary fix meanwhile? 

I think this thread adresses the same issue:

https://github.com/mongodb-labs/mongo-connector/issues/264

**Comment by behackett** [comment]
The problem was in MongoDB itself, but the solution to the problem was a new cursor API in MongoDB 3.2 that is not supported by PyMongo 2.x. There is no workaround.

**Comment by aherlihy** [comment]
@ryan-lang @phber We introduced support for PyMongo 3.x in the most recent release, so I'm going to close this issue. Please feel free to re-open or file a new issue if the problem persists. Thank you!
